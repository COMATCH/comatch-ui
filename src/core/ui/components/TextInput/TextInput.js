import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import isNil from 'lodash/isNil';
import noop from 'lodash/noop';
import { InputLabel } from '../InputLabel';
import { InputError } from '../InputError';

import './TextInput.scss';

// TextInput needs to have an implementation as an ES6 class, as opposed
// to a functional component, as it is used also internally by the DateInput.
// The DateInput uses the npm package react-datepicker, which needs to maintain
// an internal ref of the TextInput, thus a class is needed.

/**
 * An input of type `text`.
 */

export class TextInput extends Component {
    static propTypes = {
        autoComplete: PropTypes.oneOf(['on', 'off', null]),
        /**
         * Additional class names
         */
        className: PropTypes.string,
        /**
         * Fallback default for `value`
         */
        defaultValue: PropTypes.string,
        disabled: PropTypes.bool,
        display: PropTypes.oneOf(['block', 'inline']),
        /**
         * SVG icon
         */
        icon: PropTypes.node,
        /**
         * HTML id attribute
         */
        id: PropTypes.string,
        inputError: PropTypes.string,
        label: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
        maxLength: PropTypes.number,
        /**
         * Used for type="number" to define minimum allowed value
         */
        min: PropTypes.number,
        /**
         * Used for type="number" to define maximum allowed value
         */
        max: PropTypes.number,
        name: PropTypes.string.isRequired,
        onBlur: PropTypes.func,
        onChange: PropTypes.func.isRequired,
        onClick: PropTypes.func,
        onFocus: PropTypes.func,
        placeholder: PropTypes.string,
        required: PropTypes.bool,
        type: PropTypes.oneOf(['text', 'tel', 'email', 'number', 'password']),
        value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    };

    static defaultProps = {
        autoComplete: null,
        className: null,
        defaultValue: '',
        disabled: null,
        display: 'block',
        icon: null,
        id: null,
        inputError: '',
        label: '',
        maxLength: null,
        min: null,
        max: null,
        onBlur: noop,
        onClick: noop,
        onFocus: noop,
        placeholder: '',
        required: false,
        type: 'text',
        value: '',
    };

    render() {
        const {
            autoComplete,
            className,
            defaultValue,
            disabled,
            display,
            icon,
            id,
            inputError,
            label,
            maxLength,
            min,
            max,
            name,
            onBlur,
            onChange,
            onClick,
            onFocus,
            placeholder,
            required,
            type,
            value,
        } = this.props;

        const classes = classNames('Input', 'TextInput', className, display, {
            disabled,
            'has-error': inputError,
            'with-icon': icon,
        });

        return (
            <div id={id} className={classes}>
                {label && <InputLabel text={label} required={required} />}
                <input
                    autoComplete={autoComplete}
                    className="TextInput__input"
                    disabled={disabled}
                    maxLength={maxLength}
                    min={min}
                    max={max}
                    name={name}
                    onBlur={onBlur}
                    onChange={onChange}
                    onClick={onClick}
                    onFocus={onFocus}
                    placeholder={placeholder}
                    type={type}
                    value={isNil(value) ? defaultValue : value}
                />
                {icon}
                {inputError && <InputError text={inputError} />}
            </div>
        );
    }
}
