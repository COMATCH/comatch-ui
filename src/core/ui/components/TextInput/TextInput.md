
## TextInput

From [`TextInput.js`](TextInput.js)

TextInput needs to have an implementation as an ES6 class, as opposed
to a functional component, as it is used also internally by the DateInput.

The DateInput uses the npm package react-datepicker, which needs to maintain
an internal ref of the TextInput, thus a class is needed.

prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**autoComplete** | `Enum('on','off',null)` | `null` | :x: | 
**className** | `String` | `null` | :x: | 
**defaultValue** | `String` | `''` | :x: | 
**disabled** | `Boolean` | `null` | :x: | 
**display** | `Enum('block','inline')` | `'block'` | :x: | 
**icon** | `ReactNode` | `null` | :x: | 
**id** | `String` | `null` | :x: | 
**inputError** | `String` | `''` | :x: | 
**label** | `Union<ReactNode \| String>` | `''` | :x: | 
**maxLength** | `Number` | `null` | :x: | 
**name** | `String` |  | :white_check_mark: | 
**onBlur** | `Function` | `noop` | :x: | 
**onChange** | `Function` |  | :white_check_mark: | 
**onClick** | `Function` | `noop` | :x: | 
**onFocus** | `Function` | `noop` | :x: | 
**placeholder** | `String` | `''` | :x: | 
**required** | `Boolean` | `false` | :x: | 
**type** | `Enum('text','tel','email','number','password')` | `'text'` | :x: | 
**value** | `Union<String \| Number>` | `''` | :x: | 



