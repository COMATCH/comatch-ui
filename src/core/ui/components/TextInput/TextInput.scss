@import '../../../styles/variables.scss';

.TextInput {
    display: inline-block;
    margin-bottom: $co-inputs-margin-bottom;

    .InputLabel,
    &.block,
    &__input {
        display: block;
    }

    &__input {
        background-color: $co-color-white;
        border: 1px solid $co-color-mid-gray;
        border-radius: 3px;
        box-shadow: 0 0 4px 0 rgba(1, 156, 185, 0), inset 0 0 2px 0 rgba(193, 193, 193, 0.5);
        font-size: $co-font-sm;
        // sizing & whitespace
        font-size: $co-font-sm;
        padding: 9px;
        height: 40px;
        width: 100%;
        // Transition effects
        transition: box-shadow 250ms ease-out;
        transform: translateZ(0);
        will-change: box-shadow;
        // reset browser defaults
        outline: 0 !important;

        &:focus {
            box-shadow: 0 0 4px 0 rgba(1, 156, 185, 1), inset 0 0 2px 0 rgba(193, 193, 193, 0.5);
        }
    }

    // States
    &.disabled {
        .TextInput__input:hover,
        .TextInput__input {
            background-color: $co-color-light-gray;;
            color: $co-color-gray;
        }

        .TextInput__icon,
        svg {
            color: $co-color-gray;
        }
    }

    &.has-error {
        .TextInput__input {
            box-shadow: 0 0 4px 0 rgba(173, 25, 25, 1), inset 0 0 2px 0 rgba(193, 193, 193, 0.5);
        }
    }

    &.with-icon {
        position: relative;

        .TextInput__input {
            padding-right: 30px;
        }

        // icon
        .TextInput__icon,
        svg {
            width: 14px;
            height: 14px;
        }

        .TextInput__icon,
        & > svg {
            right: 10px;
            position: absolute;
            transform: translateY(-26px);
        }

        .TextInput__icon:not(.clickable),
        & > svg:not(.clickable) {
            pointer-events: none;
        }

        .TextInput__icon.clickable,
        & > svg.clickable {
            cursor: pointer;
        }
    }
}

// Defining the color of the placeholder
.TextInput::placeholder { /* Chrome, Firefox, Opera, Safari 10.1+ */
    color: $co-color-mid-gray;
    opacity: 1; /* Firefox */
}

.TextInput:-ms-input-placeholder { /* Internet Explorer 10-11 */
    color: $co-color-mid-gray;
}

.TextInput::-ms-input-placeholder { /* Microsoft Edge */
    color: $co-color-mid-gray;
}
