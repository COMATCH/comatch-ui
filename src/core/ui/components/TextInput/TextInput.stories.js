/* eslint-disable max-len */
import React, { Component } from 'react';
import Highlight from 'react-highlight';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { TextInput } from './TextInput';

const handleChange = (event) => action('changed')(event.target.name, event.target.value);

class TextInputContainer extends Component {
    state = { description: 'this is my text' };

    handleChange = (event) => {
        this.setState({ description: event.target.value });
        action('changed in container')(event.target.name, event.target.value);
    };

    handleFocus = (event) => {
        action('focused in container')(event);
    };

    handleBlur = (event) => {
        action('blured in container')(event);
    };

    render() {
        return (
            <TextInput
                name="field"
                onChange={this.handleChange}
                onFocus={this.handleFocus}
                onBlur={this.handleBlur}
                value={this.state.description}
            />
        );
    }
}

storiesOf('TextInput', module)
    .add('with basic style', () => (
        <>
            <TextInput name="field" onChange={action('changed')} />
            <Highlight className="html">{`<TextInput name="field" onChange={action('changed')} />`}</Highlight>
        </>
    ))
    .add('with placeholder', () => (
        <>
            <TextInput name="field" onChange={handleChange} placeholder="Placeholder text" />
            <Highlight className="html">{`<TextInput name="field" onChange={handleChange} placeholder="Placeholder text" />`}</Highlight>
        </>
    ))
    .add('with value', () => (
        <>
            <TextInput name="field" onChange={handleChange} value="Value text" />
            <Highlight className="html">{`<TextInput name="field" onChange={handleChange} value="Value text" />`}</Highlight>
        </>
    ))
    .add('with label', () => (
        <>
            <TextInput name="field" onChange={handleChange} label="Label text" />
            <Highlight className="html">{`<TextInput name="field" onChange={handleChange} label="Label text" />`}</Highlight>
        </>
    ))
    .add('with error', () => (
        <>
            <TextInput name="field" onChange={handleChange} label="Label text" inputError="Error text" />
            <Highlight className="html">{`<TextInput name="field" onChange={handleChange} label="Label text" inputError="Error text" />`}</Highlight>
        </>
    ))
    .add('disabled', () => (
        <>
            <TextInput name="field" onChange={handleChange} placeholder="Placeholder text" disabled />
            <Highlight className="html">{`<TextInput name="field" onChange={handleChange} placeholder="Placeholder text" disabled />`}</Highlight>
        </>
    ))
    .add('in a container', () => (
        <>
            <TextInputContainer />
            <Highlight className="javascript">{`class TextInputContainer extends Component {
    state = { description: 'this is my text' };
            
    handleChange = (event) => {
        this.setState({ description: event.target.value });
        action('changed in container')(event.target.name, event.target.value);
    };
            
    handleFocus = (event) => {
        action('focused in container')(event);
    };
            
    handleBlur = (event) => {
        action('blured in container')(event);
    };
            
    render() {
        return (
            <TextInput
            name="field"
            onChange={this.handleChange}
            onFocus={this.handleFocus}
            onBlur={this.handleBlur}
            value={this.state.description}
            />
        );
    }
}`}</Highlight>
        </>
    ));
