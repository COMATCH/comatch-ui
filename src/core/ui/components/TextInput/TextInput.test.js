import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { lorem } from 'faker';
import noop from 'lodash/noop';
import '../../../config/tests/setup';

import { TextInput } from './TextInput';

describe('TextInput', () => {
    const COMMON_PROPS = {
        name: 'text-input-unit-test',
        onChange: jest.fn(),
    };

    it('should render correctly', () => {
        const props = {
            ...COMMON_PROPS,
        };

        const thing = renderer.create(<TextInput {...props} />).toJSON();
        expect(thing).toMatchSnapshot();
    });

    it('should render with default props', () => {
        const props = {
            ...COMMON_PROPS,
        };

        const thing = mount(<TextInput {...props} />);

        // Required props:
        expect(thing.prop('name')).toEqual(props.name);
        expect(thing.prop('onChange')).toEqual(props.onChange);

        // Defaults:
        expect(thing.prop('autoComplete')).toEqual(null);
        expect(thing.prop('className')).toEqual(null);
        expect(thing.prop('defaultValue')).toEqual('');
        expect(thing.prop('disabled')).toEqual(null);
        expect(thing.prop('display')).toEqual('block');
        expect(thing.prop('icon')).toEqual(null);
        expect(thing.prop('id')).toEqual(null);
        expect(thing.prop('inputError')).toEqual('');
        expect(thing.prop('label')).toEqual('');
        expect(thing.prop('maxLength')).toEqual(null);
        expect(thing.prop('onBlur')).toEqual(noop);
        expect(thing.prop('onClick')).toEqual(noop);
        expect(thing.prop('onFocus')).toEqual(noop);
        expect(thing.prop('placeholder')).toEqual('');
        expect(thing.prop('required')).toEqual(false);
        expect(thing.prop('type')).toEqual('text');
        expect(thing.prop('value')).toEqual('');
    });

    it('should render with certain props', () => {
        const props = {
            ...COMMON_PROPS,

            className: lorem.word(),
            defaultValue: lorem.word(),
            id: lorem.word(),
            inputError: lorem.word(),
            label: lorem.word(),
            name: lorem.word(),
            placeholder: lorem.word(),
            value: lorem.word(),

            disabled: false,
            required: true,

            autoComplete: 'on',
            display: 'inline',
            type: 'password',

            maxLength: 100,

            icon: <i />,

            onBlur: jest.fn(),
            onClick: jest.fn(),
            onFocus: jest.fn(),
        };

        const thing = mount(<TextInput {...props} />);

        // Required Props:
        expect(thing.prop('name')).toEqual(props.name);
        expect(thing.prop('onChange')).toEqual(props.onChange);

        // Optional Props:
        expect(thing.prop('className')).toEqual(props.className);
        expect(thing.prop('defaultValue')).toEqual(props.defaultValue);
        expect(thing.prop('id')).toEqual(props.id);
        expect(thing.prop('inputError')).toEqual(props.inputError);
        expect(thing.prop('label')).toEqual(props.label);
        expect(thing.prop('name')).toEqual(props.name);
        expect(thing.prop('placeholder')).toEqual(props.placeholder);
        expect(thing.prop('value')).toEqual(props.value);

        expect(thing.prop('disabled')).toEqual(props.disabled);
        expect(thing.prop('required')).toEqual(props.required);

        expect(thing.prop('autoComplete')).toEqual(props.autoComplete);
        expect(thing.prop('display')).toEqual(props.display);
        expect(thing.prop('type')).toEqual(props.type);

        expect(thing.prop('maxLength')).toEqual(props.maxLength);

        expect(thing.prop('icon')).toEqual(props.icon);

        expect(thing.prop('onBlur')).toEqual(props.onBlur);
        expect(thing.prop('onClick')).toEqual(props.onClick);
        expect(thing.prop('onFocus')).toEqual(props.onFocus);
    });
});
