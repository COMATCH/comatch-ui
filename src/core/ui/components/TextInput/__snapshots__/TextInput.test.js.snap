// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput should render correctly 1`] = `
<div
  className="Input TextInput block"
  id={null}
>
  
  <input
    autoComplete={null}
    className="TextInput__input"
    disabled={null}
    max={null}
    maxLength={null}
    min={null}
    name="text-input-unit-test"
    onBlur={[Function]}
    onChange={[MockFunction]}
    onClick={[Function]}
    onFocus={[Function]}
    placeholder=""
    type="text"
    value=""
  />
  
</div>
`;

exports[`TextInput should render correctly with a Label 1`] = `
<div
  className="Input TextInput block"
  id={null}
>
  <label
    className="InputLabel"
    htmlFor={null}
    id={null}
  >
    Testing Label
  </label>
  <input
    autoComplete={null}
    className="TextInput__input"
    disabled={null}
    max={null}
    maxLength={null}
    min={null}
    name="text-input-unit-test"
    onBlur={[Function]}
    onChange={[MockFunction]}
    onClick={[Function]}
    onFocus={[Function]}
    placeholder=""
    type="text"
    value=""
  />
  
</div>
`;

exports[`TextInput should render correctly with an Icon 1`] = `
<div
  className="Input TextInput block with-icon"
  id={null}
>
  
  <input
    autoComplete={null}
    className="TextInput__input"
    disabled={null}
    max={null}
    maxLength={null}
    min={null}
    name="text-input-unit-test"
    onBlur={[Function]}
    onChange={[MockFunction]}
    onClick={[Function]}
    onFocus={[Function]}
    placeholder=""
    type="text"
    value=""
  />
  <i />
  
</div>
`;

exports[`TextInput should render correctly with an InputError 1`] = `
<div
  className="Input TextInput block has-error"
  id={null}
>
  
  <input
    autoComplete={null}
    className="TextInput__input"
    disabled={null}
    max={null}
    maxLength={null}
    min={null}
    name="text-input-unit-test"
    onBlur={[Function]}
    onChange={[MockFunction]}
    onClick={[Function]}
    onFocus={[Function]}
    placeholder=""
    type="text"
    value=""
  />
  <span
    className="InputError"
  >
    Testing InputError Message
  </span>
</div>
`;
